tests/unit/remote/test_remote.py

```diff
@@ -25,3 +27,14 @@ def test_remote_without_checksum_jobs_default(dvc):
 
     remote = Remote(dvc, name="without_checksum_jobs")
     assert remote.checksum_jobs == remote.CHECKSUM_JOBS
+
+
+@pytest.mark.parametrize("remote_cls", [RemoteGS, RemoteS3])
+def test_makedirs_not_create_for_top_level_path(remote_cls, mocker):
+    url = "{.scheme}://bucket/".format(remote_cls)
+    remote = remote_cls(None, {"url": url})
+    exc = Exception("should not try to create dir on top-level path")
+    client = mocker.PropertyMock(side_effect=exc)
+    # we use remote clients with same name as scheme to interact with remote
+    setattr(remote_cls, remote_cls.scheme, client)
```

Comment:
In one of his PR's @Suor proposed approach utilizing `funcy`:
`mocker.patch.object(remote_cls, remote_cls.scheme, property(raiser(Exception("should not try to create dir on top-level path"))))` .

We might want to consider this approach, but its a matter of taste, so don't consider it as a change request.

Reply:
Why not just mock and `assert not mock.called`? Is there a reason we try to use setattr hack instead of simply using regular mocker facilities?

Reply:
Should be good now. :)